tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts

```diff
@@ -504,8 +508,81 @@ const reducer = createReducer(
             lastLoadedTimeInMs: Date.now(),
           },
           sourceFileList: sourceFileList.sourceFiles,
+          fileContents: [...state.sourceCode.fileContents],
         },
       };
+      const newNumFiles = sourceFileList.sourceFiles.length;
+      const oldNumFiles = newState.sourceCode.fileContents.length;
+      if (newNumFiles > oldNumFiles) {
+        newState.sourceCode.fileContents.push(
+          ...new Array<SourceFileContent>(newNumFiles - oldNumFiles).fill({
+            loadState: DataLoadState.NOT_LOADED,
+            lines: null,
+          })
+        );
+      }
+      return newState;
+    }
+  ),
+  on(
+    actions.sourceFileRequested,
+    (state: DebuggerState, sourceFileSpec: SourceFileSpec): DebuggerState => {
+      const newState: DebuggerState = {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          fileContents: state.sourceCode.fileContents.slice(),
+        },
+      };
+      const fileIndex = newState.sourceCode.sourceFileList.findIndex(
+        (fileSpec) =>
+          fileSpec.host_name === sourceFileSpec.host_name &&
+          fileSpec.file_path === sourceFileSpec.file_path
+      );
+      if (fileIndex >= 0) {
+        newState.sourceCode.fileContents[fileIndex].loadState =
+          DataLoadState.LOADING;
+      } else {
+        throw Error(
+          `Cannot find the following file in file list: ` +
+            `host_name="${sourceFileSpec.host_name}", ` +
+            `file_path="${sourceFileSpec.file_path}"`
+        );
+      }
+      return newState;
+    }
+  ),
+  on(
+    actions.sourceFileLoaded,
+    (
+      state: DebuggerState,
+      sourceFileResponse: SourceFileResponse
+    ): DebuggerState => {
+      const newState: DebuggerState = {
+        ...state,
+        sourceCode: {
+          ...state.sourceCode,
+          fileContents: state.sourceCode.fileContents.slice(),
+        },
+      };
+      const fileIndex = newState.sourceCode.sourceFileList.findIndex(
+        (fileSpec) =>
+          fileSpec.host_name === sourceFileResponse.host_name &&
+          fileSpec.file_path === sourceFileResponse.file_path
+      );
+      if (fileIndex >= 0) {
```

Comment:
store_utils looks nice.  Optional, but wdyt about adding `getFileContents(debuggerState, sourceFileSpec)` to the `store_utils.ts`?  Both `sourceFileSpec` and `fileIndex` act like a file `Id`, but the latter happens to be implemented in a way that forces clients to know that fileList and fileContents are linked.

Reply:
I considered `getFileContents(debuggerState, sourceFileSpec)` but decided against it, for the following reasons:

1. It's nicer to have the utility functions decoupled from the details of the state object, which has a large and complex structure.
2. `getFileContents(debuggerState, sourceFileSpec)` is harder to use from the effects module (`debugger_effects.ts`), because in that module we don't access the entire state. We access pieces of the state via selectors instead.